main/src/ca/uwaterloo/flix/language/phase/Optimizer.scala

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright 2015-2016 Magnus Madsen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.language.ast.SimplifiedAst.Definition.Constant
+import ca.uwaterloo.flix.language.ast.SimplifiedAst.Expression._
+import ca.uwaterloo.flix.language.ast.SimplifiedAst.{Expression, ExpressionFolder, FreeVar, Root}
+import ca.uwaterloo.flix.language.ast.{BinaryOperator, Name, SourceLocation, Type}
+
+object Optimizer {
+
```

Comment:
I will just add some more ideas here:

- Fold constants, e.g. 1 + 1 => 2
- Eliminate checks for unary enums (pretty similar to the one for unit)
- Eliminate unnecessary allocations, e.g. (1, 2) match { case (1, 2) => } this may be hard (?)